### PR TITLE
LibJS: Align array splice with spec

### PR DESCRIPTION
**Pull request by flostellbrink**
Added the spec copy for array splice to the method.

Tweaked variable names and logic to align closer with the specs.

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 1]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
+    // FIXME: Slightly different from spec logic
+    // 5. Else if relativeStart < 0, let actualStart be max(len + relativeStart, 0).
     if (relative_start < 0)
         actual_start = max((ssize_t)initial_length + relative_start, (ssize_t)0);
+    // 6. Else, let actualStart be min(relativeStart, len).
     else
         actual_start = min(relative_start, initial_length);
 
+    // 7. Let itemCount be the number of elements in items.
+    // REMARK: 0 if no items, updated later if there are items
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: We call these NOTE, but this also seems fairly obvious from the name

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 2]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
 
+    // 9. Else if deleteCount is not present, then
     if (vm.argument_count() == 1) {
+        // a. Let actualDeleteCount be len - actualStart.
         actual_delete_count = initial_length - actual_start;
-    } else if (vm.argument_count() >= 2) {
+    }
+    // 10. Else,
+    else if (vm.argument_count() >= 2) {
+        // REMARK: Handles 7. if items are present
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: NOTE

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 3]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
         insert_count = vm.argument_count() - 2;
+
+        // a. Let dc be ? ToIntegerOrInfinity(deleteCount).
         auto delete_count = TRY(vm.argument(1).to_integer_or_infinity(vm));
+
+        // b. Let actualDeleteCount be the result of clamping dc between 0 and len - actualStart.
+        // FIXME there's gotta be an actual clamp function somewhere
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: FIXME: with a colon, same below

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 4]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
+    // FIXME loop index should be k
+    for (u64 i = 0; i < insert_count; ++i) {
+        // a. Perform ? Set(O, ! ToString(𝔽(k)), E, true).
         TRY(this_object->set(actual_start + i, vm.argument(i + 2), Object::ShouldThrowExceptions::Yes));
 
+        // b. Set k to k + 1.
+    }
+
+    // 20. Perform ? Set(O, "length", 𝔽(len - actualDeleteCount + itemCount), true).
+    // FIXME not spec logic
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: ?

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 5]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
+
+            // v. Set k to k - 1.
         }
     }
 
-    for (u64 i = 0; i < insert_count; ++i)
+    // 18. Set k to actualStart.
+    // 19. For each element E of items, do
+    // FIXME loop index should start at actual_start
+    // FIXME loop index should be k
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: we can't range-for loop over `items` here so I don't think these FIXMEs make sense

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 6]
```diff
@@ -1565,34 +1563,32 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::splice)
...
 
             // ii. Set k to k - 1.
         }
     }
     // 17. Else if itemCount > actualDeleteCount, then
-    else if (insert_count > actual_delete_count) {
+    else if (item_count > actual_delete_count) {
         // a. Set k to (len - actualDeleteCount).
         // b. Repeat, while k > actualStart,
         // FIXME: loop index should be k
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: forgot to remove the fixme

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 7]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
+            // ii. Let to be ! ToString(𝔽(k + itemCount)).
+            auto to = i + insert_count;
+
+            // iii. If ? HasProperty(O, from) is true, then
             if (TRY(this_object->has_property(from))) {
+                // 1. Let fromValue be ? Get(O, from).
                 auto from_value = TRY(this_object->get(from));
+
+                // 2. Perform ? Set(O, to, fromValue, true).
+                // FIXME: O != to
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: ?

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 8]
```diff
@@ -1445,87 +1445,184 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::sort)
...
                 TRY(this_object->set(to, from_value, Object::ShouldThrowExceptions::Yes));
-            } else {
+            }
+            // iv. Else,
+            else {
+                // 1. Perform ? DeletePropertyOrThrow(O, to).
+                // FIXME: O != to
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: ?

**Review by linusg — CHANGES_REQUESTED** [review]
Looks nice, thanks, but a few of these FIXMES need fixing :)

**Comment by flostellbrink** [comment]
Thanks for the review @linusg 
I only added annotations and notes in the first commit to make it very clear what the next two commits addressed.
Seems like you guys tend to prefer squashing everything up, gonna go a head and do that now :)

File: Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp (hunk 1 above)
Comment (flostellbrink) [reply Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp]: Good to know :)

**Comment by linusg** [comment]
This broke 11 test262 tests, so I have reverted it again. Feel free to fix those and PR again:

```
    test/built-ins/Array/prototype/splice/15.4.4.12-9-c-ii-1.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.1_T4.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.1_T6.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.2_T4.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.3_T4.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.3_T5.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.4_T4.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A1.4_T6.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A4_T2.js                     ✅ -> ❌
    test/built-ins/Array/prototype/splice/S15.4.4.12_A6.1_T1.js                   ✅ -> ❌
    test/built-ins/Array/prototype/splice/length-near-integer-limit-grow-array.js ✅ -> ❌
```

**Pull request merged**